tasks/main.yml

```diff
@@ -32,29 +32,88 @@
   tags:
     - keepalived-config
 
-- name: install keepalived
-  package:
-    name: "{{ keepalived_package_name }}"
-    state: "{{ keepalived_package_state }}"
-    update_cache: "{{ (ansible_pkg_mgr == 'apt') | ternary('yes', omit) }}"
-    cache_valid_time: "{{ (ansible_pkg_mgr == 'apt') | ternary(cache_timeout, omit) }}"
-  tags:
-    - keepalived-apt-packages
-    - keepalived-packages
-  notify:
-    - restart keepalived
+- name: Install keepalived with packages
+  block:
+    - name: install keepalived
+      package:
+        name: "{{ keepalived_package_name }}"
+        state: "{{ keepalived_package_state }}"
+        update_cache: "{{ (ansible_pkg_mgr == 'apt') | ternary('yes', omit) }}"
+        cache_valid_time: "{{ (ansible_pkg_mgr == 'apt') | ternary(cache_timeout, omit) }}"
+      tags:
+        - keepalived-apt-packages
+        - keepalived-packages
+      notify:
+        - restart keepalived
 
-- name: Ensure no new "samples" folder appeared
-  file:
-    path: /etc/keepalived/samples/
-    state: absent
+    - name: Ensure no new "samples" folder appeared
+      file:
+        path: /etc/keepalived/samples/
+        state: absent
+      when:
+        - ansible_os_family | lower == 'debian'
   when:
-    - ansible_os_family | lower == 'debian'
+    - not keepalived_snap | bool
+
+- name: Install keepalived with snap
+  block:
+    - name: Install snapd and fuse
+      package:
+        name: "{{ item }}"
+        state: "{{ keepalived_package_state }}"
+        update_cache: "{{ (ansible_pkg_mgr == 'apt') | ternary('yes', omit) }}"
+        cache_valid_time: "{{ (ansible_pkg_mgr == 'apt') | ternary(cache_timeout, omit) }}"
+      loop:
+        - snapd
+        - fuse
+
+    - name: Ensure snapd.service is enabled and running
+      ansible.builtin.systemd:
+        name: snapd.service
+        enabled: true
+        masked: false
+        state: started
+
+    - name: Ensure snap links
+      ansible.builtin.file:
+        src: /var/lib/snapd/snap
+        dest: /snap
+        owner: root
+        group: root
+        state: link
+      when:
+        - ansible_facts["os_family"] == "RedHat"
+
+    - name: Ensure snap core
+      community.general.snap:
+        name: core
+      # snap core may result in a weird error that resolves when repeating the action
+      retries: 3
+      register: snapd_core
+      until: snapd_core is succeeded
+
+    - name: Install keepalived from snap
+      community.general.snap:
+        name: keepalived
+        classic: true
+        channel: "{{ keepalived_snap_channel }}"
+
+    - name: Disable snap daemon
+      command: snap stop --disable keepalived
+      become: true
+      changed_when: false
```

Comment:
I am reluctant to have a changed_when false where there is possibly a change. Is there no module that does this more appropriately, keeping idempotency and real status reporting?

Reply:
The module seems a bit old and simply runs commands remotely. I was unable to find something else.
I added an attempt to interact with the `snap` API with https://github.com/evrardjp/ansible-keepalived/pull/161/commits/ff2d15d9f9a8002be04fea57c7f75826ab67e0a5.

Reply:
Sorry for any possible back and forth here. I supposed the module of snap is keeping idempotency and check mode.
Do you mean that the module isn't perfect and don't really do that?

If you're telling me the module is not perfect, I would prefer that we document _why_ we don't use it, so that we can re-evaluate when it's improved.

Reply:
The module just installs and uninstalls snaps.
Some snaps, like this one, have "services" which are defined by the dev, which in the case of `keepalived` simply runs it with no particular frills. A service file is the better choice.
To my knowledge there is no module that stops/disables the snap daemon, which is required to use a custom service file.